lib/location/method_gnss.c

```diff
@@ -828,13 +906,135 @@ static void method_gnss_positioning_work_fn(struct k_work *work)
+		if (agps_expiry->ephe_expiry[i] <= ephe_expiry_threshold) {
+			expired_ephes++;
+		}
+
+		if (agps_expiry->alm_expiry[i] <= AGPS_EXPIRY_THRESHOLD) {
+			expired_alms++;
+		}
+	}
+
+	if (IS_ENABLED(CONFIG_NRF_CLOUD_PGPS)) {
+		expired_ephes_min_count = PGPS_EPHE_MIN_COUNT;
+	} else {
+		expired_ephes_min_count = AGPS_EPHE_MIN_COUNT;
+	}
+
+	if (expired_ephes >= expired_ephes_min_count) {
+		agps_request.sv_mask_ephe = 0xffffffff;
+	}
+
+	if (expired_alms >= AGPS_ALM_MIN_COUNT) {
+		agps_request.sv_mask_alm = 0xffffffff;
+	}
+
+	if (agps_expiry->utc_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_GPS_UTC_REQUEST;
+	}
+
+	if (agps_expiry->klob_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_KLOBUCHAR_REQUEST;
+	}
+
+	if (agps_expiry->data_flags & NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_SYS_TIME_AND_SV_TOW_REQUEST;
+	}
+
+	if (agps_expiry->integrity_expiry <= AGPS_EXPIRY_THRESHOLD) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_INTEGRITY_REQUEST;
+	}
+
+	if (agps_expiry->data_flags & NRF_MODEM_GNSS_AGPS_POSITION_REQUEST) {
+		agps_request.data_flags |= NRF_MODEM_GNSS_AGPS_POSITION_REQUEST;
+	}
+
+	LOG_DBG("Expired ephemerides: %d, almanacs: %d", expired_ephes, expired_alms);
+
+	LOG_DBG("Assistance request sv_mask_ephe: 0x%08x, sv_mask_alm: 0x%08x, data_flags: 0x%02x",
+		agps_request.sv_mask_ephe, agps_request.sv_mask_alm, agps_request.data_flags);
+
+	return agps_request.sv_mask_ephe != 0x0 ||
+	       agps_request.sv_mask_alm != 0x0 ||
+	       agps_request.data_flags != 0x0;
+}
+
+/* Queries assistance data need from GNSS.
+ *
+ * To maintain backward compatibility with older modem firmware versions, two different methods are
+ * supported. If the nrf_modem_gnss_agps_expiry_get() API is not supported (pre-v1.3.2 modem
+ * firmware), GNSS is briefly started to trigger the NRF_MODEM_GNSS_EVT_AGPS_REQ event from GNSS
+ * in case assistance data is currently needed.
+ */
 static void method_gnss_agps_req_work_fn(struct k_work *work)
 {
-	/* Start and stop GNSS to trigger NRF_MODEM_GNSS_EVT_AGPS_REQ event if assistance data
-	 * is needed.
-	 */
-	nrf_modem_gnss_start();
-	nrf_modem_gnss_stop();
+	int err;
+	struct nrf_modem_gnss_agps_expiry agps_expiry;
+
+	err = nrf_modem_gnss_agps_expiry_get(&agps_expiry);
+	if (err) {
+		if (err == -NRF_EOPNOTSUPP) {
+			LOG_DBG("nrf_modem_gnss_agps_expiry_get() not supported by the modem "
+				"firmware");
+
+			/* Start and stop GNSS to trigger NRF_MODEM_GNSS_EVT_AGPS_REQ event if
+			 * assistance data is needed.
+			 */
+			nrf_modem_gnss_start();
+			nrf_modem_gnss_stop();
+		} else {
+			LOG_WRN("nrf_modem_gnss_agps_expiry_get() failed, err %d", err);
+		}
+
+		return;
+	}
+
+	agps_expiry_get_supported = true;
+
+	if (method_gnss_agps_expiry_process(&agps_expiry)) {
+		method_gnss_assistance_request();
+	}
```

Comment:
Cannot use proper line but check whether starting and stopping GNSS is valid in comments on line 1083:
```
	 * downloaded before GNSS is started. GNSS is briefly started and stopped to trigger
	 * the NRF_MODEM_GNSS_EVT_AGPS_REQ event, which in turn causes the A-GPS data download
```

Reply:
Yes, it's valid if `nrf_modem_gnss_agps_expiry_get()` is not supported. I'll mention this in the comment for clarity.

Reply:
It came to my mind, that we have a flag for `nrf_modem_gnss_agps_expiry_get()` support. Thus, the sleep could be done only when needed. However, the flag is set only when `method_gnss_agps_req_work` is executed for the first time, so it has not yet been set when execution gets here for the first time. Maybe we can just always sleep for 100 ms here. 😄

Reply:
Yep, sleeping 100ms always is fine